models/ctsm/R/write_params-clm.R

```diff
@@ -1,670 +1,101 @@
+# TODO: update to allow it to pick between CLM4.5 and CLM5 (CLM5 == CTSM) parameter set based on refcase, user selection
+# TODO: match Pecan pft names with CTSM pft names when Pecan name is not identical to a CTSM name
+## See issue https://github.com/PecanProject/pecan/issues/1008
 
+write_params_ctsm <-
+  function(defaults = system.file('clm5_params.c171117_0001.nc', package = 'PEcAn.CTSM'),
+           trait.values,
+           settings,
+           run.id = 1, 
+           local.rundir = tempdir()) {
+    
+    ## Copy and open default parameter files
+    ctsm.param.default <-
+      system.file('clm5_params.c171117_0001.nc', package = 'PEcAn.CTSM')
+    ctsm.param.file <-
+      file.path(local.rundir, paste0("ctsm_params.", run.id, ".nc"))
+    file.copy(ctsm.param.default, ctsm.param.file)
+    ctsm.param.nc <- ncdf4::nc_open(ctsm.param.file, write = TRUE)
+    
+    ## Loop over PFTS
+    npft <- length(trait.values)
+    PEcAn.logger::logger.debug('there are ',
+                               npft,
+                               'PFTs in this run, they are named:',
+                               names(trait.values))
+    ctsm_pftnames <-
+      trimws(tolower(ncdf4::ncvar_get(ctsm.param.nc, "pftname")))
+    for (i in seq_len(npft)) {
+      pft.name <- names(trait.values)[[i]]
+      if (is.null(pft.name) || is.na(pft.name)) {
+        PEcAn.logger::logger.error("pft.name missing")
+      } else {
+        PEcAn.logger::logger.info(paste("PFT =", pft.name))
+        PEcAn.logger::logger.debug(paste0("ctsm PFT number: ", which(ctsm_pftnames == pft.name)))
+      }
+      if (pft.name == 'env')
+        next   ## HACK, need to remove env from default
+      
+      ## Match PFT name to COLUMN
+      ipft <- match(tolower(pft.name), ctsm_pftnames)
+      PEcAn.logger::logger.debug(paste0("CTSM pft index number: ", ipft))
+      
+      if (is.na(ipft)) {
+        PEcAn.logger::logger.severe(
+          paste(
+            "Unmatched PFT",
+            pft.name,
+            "in CTSM PEcAn does not yet support non-default PFTs for this model"
+          )
+        )
+      }
+      
+      ## Special variables used in conversions
+      leafC <- 0.48
+      fnr <- 7.16 # mass ratio of total Rubisco molecular mass to nitrogen in Rubisco (g Rubisco g-1 N in Rubisco)
+      ar <- 60 # specific activity of Rubisco (µmol CO2 g-1 Rubisco s-1)
+      
+      ## Loop over VARIABLES
+      pft_pecan_vals <- trait.values[[i]]
+      for (v in seq_along(pft_pecan_vals)) {
+        pecan_var <- names(pft_pecan_vals)[v]
 
-# write_params_fates ------------------------------------------------------
-write_params_fates <- function(defaults = system.file('????', package = 'PEcAn.FATES'),
-                              trait.values, settings, run.id){
+        ### ----- Leaf physiological parameters
+        update_vars <- function(nc = ctsm.param.nc, ctsm_var, ctsm_vals, start = ipft, count = 1){
```

Comment:
In general, it's a good idea to stick function arguments with default values after all of the functions without defaults. That will also allow (if you want) to remove a bunch of the function argument names in the block below (i.e. you can just call `update_vars("slatop", udunits2::ud.convert(...))`.